pkg/share/manager/cs3/cs3.go

```diff
@@ -0,0 +1,564 @@
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	collaboration "github.com/cs3org/go-cs3apis/cs3/sharing/collaboration/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	typespb "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	ctxpkg "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/cs3org/reva/v2/pkg/errtypes"
+	"github.com/cs3org/reva/v2/pkg/share"
+	"github.com/cs3org/reva/v2/pkg/share/manager/registry"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/indexer"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/indexer/option"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/metadata"
+	"github.com/cs3org/reva/v2/pkg/utils"
+	"github.com/google/uuid"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"google.golang.org/genproto/protobuf/field_mask"
+)
+
+//go:generate mockery -name Storage
+//go:generate mockery -name Indexer
+
+// Storage is the interface to the metadata storage backend
+type Storage interface {
+	metadata.Storage
+}
+
+// Indexer is the interface to the indexer being used for indexing shares
+type Indexer interface {
+	AddIndex(t interface{}, indexBy option.IndexBy, pkName, entityDirName, indexType string, bound *option.Bound, caseInsensitive bool) error
+	Add(t interface{}) ([]indexer.IdxAddResult, error)
+	FindBy(t interface{}, field string, val string) ([]string, error)
+	Delete(t interface{}) error
+}
+
+// Manager implements a share manager using a cs3 storage backend
+type Manager struct {
+	storage Storage
+	indexer Indexer
+
+	initialized bool
+}
+
+// ReceivedShareMetadata hold the state information or a received share
+type ReceivedShareMetadata struct {
+	State      collaboration.ShareState `json:"state"`
+	MountPoint *provider.Reference      `json:"mountpoint"`
+}
+
+func init() {
+	registry.Register("cs3", NewDefault)
+}
+
+type config struct {
+	GatewayAddr       string `mapstructure:"gateway_addr"`
+	ProviderAddr      string `mapstructure:"provider_addr"`
+	ServiceUserID     string `mapstructure:"service_user_id"`
+	ServiceUserIdp    string `mapstructure:"service_user_idp"`
+	MachineAuthAPIKey string `mapstructure:"machine_auth_apikey"`
+}
+
+// NewDefault returns a new manager instance with default dependencies
+func NewDefault(m map[string]interface{}) (share.Manager, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		err = errors.Wrap(err, "error creating a new manager")
+		return nil, err
+	}
+
+	s, err := metadata.NewCS3Storage(c.GatewayAddr, c.ProviderAddr, c.ServiceUserID, c.ServiceUserIdp, c.MachineAuthAPIKey)
+	if err != nil {
+		return nil, err
+	}
+	indexer := indexer.CreateIndexer(s)
+
+	return New(s, indexer)
+}
+
+// New returns a new manager instance
+func New(s Storage, indexer Indexer) (*Manager, error) {
+	return &Manager{
+		storage:     s,
+		indexer:     indexer,
+		initialized: false,
+	}, nil
+}
+
+func (m *Manager) initialize() error {
+	if m.initialized {
```

Comment:
We can have an concurrency issue here. Imagine two routines go into this function at the same the time. `m.initialized` would be false for both, so both start the initialization process. This can lead to unforeseeable consequences

My best approach was to `Lock` if not initialized and then check for `initialized` again. But there might be better ones.

Reply:
My idea was that running the initializing steps multiple times should be fine because they should be idempotent. I also *think* that running them in parallel shouldn't do any harm, but I agree it's better to err on the safe side here. I'll add some locking to it.